sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/SaveAsHiveFile.scala

```diff
@@ -97,7 +99,34 @@ private[hive] trait SaveAsHiveFile extends DataWritingCommand {
       options = Map.empty)
   }
 
-  protected def getExternalTmpPath(
+  // Mostly copied from Context.java#getMRTmpPath of Hive 2.3.
+  // Visible for testing.
+  private[execution] def getMRTmpPath(
+      hadoopConf: Configuration,
+      sessionScratchDir: String,
+      scratchDir: String): Path = {
+
+    // Hive's getMRTmpPath uses nonLocalScratchPath + '-mr-10000',
+    // which is ruled by 'hive.exec.scratchdir' including file system.
+    // This is the same as Spark's #oldVersionExternalTempPath.
+    // Only difference between #oldVersionExternalTempPath and Hive 2.3.0's is HIVE-7090.
+    // HIVE-7090 added user_name/session_id on top of 'hive.exec.scratchdir'
+    // Here it uses session_path unless it's emtpy, otherwise uses scratchDir.
+    val sessionPath = if (!sessionScratchDir.isEmpty) sessionScratchDir else scratchDir
+    val mrScratchDir = oldVersionExternalTempPath(new Path(sessionPath), hadoopConf, sessionPath)
```

Comment:
Any benefit to use sessionScratchDir here except for following Hive?

Reply:
Several HDFS scratch directories are created during start SessionState.
If the session scratch directory is created in the path specified in `_hive.hdfs.session.path`, the directory should be used. If it is not specified, then we just use scratchDir.